x/ibc/02-client/keeper/client.go

```diff
@@ -73,23 +75,38 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, header exported.H
 
 // CheckMisbehaviourAndUpdateState checks for client misbehaviour and freezes the
 // client if so.
-func (k Keeper) CheckMisbehaviourAndUpdateState(ctx sdk.Context, clientID string, evidence exported.Evidence) error {
-	clientState, found := k.GetClientState(ctx, clientID)
+//
+// NOTE: In the first implementation, only Tendermint misbehaviour evidence is
+// supported.
+func (k Keeper) CheckMisbehaviourAndUpdateState(ctx sdk.Context, evidence evidenceexported.Evidence) error {
```

Comment:
This doesn't make sense to me. My understanding is that the client should be frozen if the entire chain breaks the consensus protocol. For example:
1. ChainA is running light client of ChainB
2. ChainB validator set commits 2 different blocks for the same height (violation of consensus protocol)
3. Evidence of this violation is submitted to ChainA
4. ChainA freezes light client of ChainB

The code here looks like it will freeze the client as soon as someone posts evidence of a single validator double signing. This doesn't make sense since it won't disrupt consensus, so long as validator power < 1/3 of total.

Thus, with code as is, a single slash event of a double-signing validator on chainB (which is still considered honest execution of Byzantine consensus protocol) will cause IBC client to freeze. Thus, a chainA no longer has any Byzantine tolerance for chainB; which doesn't seem like the intended behavior from looking at spec

@cwgoes

Reply:
I believe the Evidence should be changed from embedding a single `tendermint.DoubleVoteEvidence` to including two different block headers that got committed by ChainB's validatorSet but are for same height

Reply:
Haha, i think i was the one to commit this mistake to the code in the first place 😆. Will put up PR for proposed fix, once i get an ACK that the problem i point out is correct

Reply:
https://github.com/cosmos/cosmos-sdk/blob/ibc-alpha/x/ibc/02-client/types/tendermint/misbehaviour.go#L22

Reply:
Further thinking here after prototyping solution:

1. The `Evidence` interface defined in `x/evidence` isn't well suited to use here, since that is Evidence against a single malicious validator. Whereas I believe the Evidence that should freeze a ICS-02 client is evidence of a malicious chain. We should have some other primitive defined in x/ibc to deal with malicious chains.

2. The overloading of the term Evidence to describe evidence against validators and evidence against chains is confusing. I think we should have a different word to talk about malicious chains and the associated proofs, (`Violation` maybe)?

Proposals: 

A) evidence module is adjusted to handle both malicious-validator evidence and malicious-chain evidence

B) malicious-chain evidence types and functions are implemented within `x/ibc` since that is the only place they will be used.

I'm partial to proposal B myself and removing any dependence of `x/evidence` from `x/ibc` for now, but interested to hear what others think. Note: validator-specific evidence isn't relevant in IBC for version 1.0.0, but will later be useful once we implement shared-security across chains.

Reply:
> This doesn't make sense to me. My understanding is that the client should be frozen if the entire chain breaks the consensus protocol. For example:
> 
> 1. ChainA is running light client of ChainB
> 2. ChainB validator set commits 2 different blocks for the same height (violation of consensus protocol)
> 3. Evidence of this violation is submitted to ChainA
> 4. ChainA freezes light client of ChainB
> 
> The code here looks like it will freeze the client as soon as someone posts evidence of a single validator double signing. This doesn't make sense since it won't disrupt consensus, so long as validator power < 1/3 of total.
> 
> Thus, with code as is, a single slash event of a double-signing validator on chainB (which is still considered honest execution of Byzantine consensus protocol) will cause IBC client to freeze. Thus, a chainA no longer has any Byzantine tolerance for chainB; which doesn't seem like the intended behavior from looking at spec
> 
> @cwgoes

Yes, this is right. Although it is also possible to choose thresholds other than 1/3 (perhaps gaining some extra precautionary safety at the expense of liveness), we should initially freeze the client if and only if two valid headers were committed at the same height.

Reply:
> Further thinking here after prototyping solution:
> 
> 1. The `Evidence` interface defined in `x/evidence` isn't well suited to use here, since that is Evidence against a single malicious validator. Whereas I believe the Evidence that should freeze a ICS-02 client is evidence of a malicious chain. We should have some other primitive defined in x/ibc to deal with malicious chains.
> 2. The overloading of the term Evidence to describe evidence against validators and evidence against chains is confusing. I think we should have a different word to talk about malicious chains and the associated proofs, (`Violation` maybe)?
> 
> Proposals:
> 
> A) evidence module is adjusted to handle both malicious-validator evidence and malicious-chain evidence
> 
> B) malicious-chain evidence types and functions are implemented within `x/ibc` since that is the only place they will be used.
> 
> I'm partial to proposal B myself and removing any dependence of `x/evidence` from `x/ibc` for now, but interested to hear what others think. Note: validator-specific evidence isn't relevant in IBC for version 1.0.0, but will later be useful once we implement shared-security across chains.

A new term might be a helpful differentiator. Malicious-chain evidence (or violation) types will be used by both the IBC on-chain light clients and Tendermint light clients (e.g. if you run `gaiacli` in proper verification mode), so we should find a way to share the definitions between those two, ideally.

Reply:
> I believe the Evidence should be changed from embedding a single tendermint.DoubleVoteEvidence to including two different block headers that got committed by ChainB's validatorSet but are for same height

yeah, that's how it is defined on the spec:

```typescript
interface Evidence {
  h1: Header
  h2: Header
}
```

Reply:
> The Evidence interface defined in x/evidence isn't well suited to use here, since that is Evidence against a single malicious validator. Whereas I believe the Evidence that should freeze a ICS-02 client is evidence of a malicious chain. We should have some other primitive defined in x/ibc to deal with malicious chains.

> A) evidence module is adjusted to handle both malicious-validator evidence and malicious-chain evidence

I think we should implement a general Evidence interface that only contains the following methods:

```go
// Evidence defines the contract which concrete evidence types of misbehavior
// must implement.
type Evidence interface {
	Route() string
	Type() string
	String() string
	Hash() cmn.HexBytes
	ValidateBasic() error
}
```

Probably the byzantine validator related methods from the current `Evidence` interface should be abstracted into another interface: 

```go
        // The consensus address of the malicious validator at time of infraction
	GetConsensusAddress() sdk.ConsAddress

	// Height at which the infraction occurred
	GetHeight() int64

	// The total power of the malicious validator at time of infraction
	GetValidatorPower() int64

	// The total validator set power at time of infraction
	GetTotalPower() int64
```

And ICS02 should implement the Evidence interface from the spec (2 headers) as well as the general interface

Reply:
> The overloading of the term `Evidence` to describe evidence against validators and evidence against chains is confusing. I think we should have a different word to talk about malicious chains and the associated proofs, (`Violation` maybe)?

Agree, ICS02 introduces the term of a client `Misbehaviour`, which is what I implemented in this PR

Reply:
> > The Evidence interface defined in x/evidence isn't well suited to use here, since that is Evidence against a single malicious validator. Whereas I believe the Evidence that should freeze a ICS-02 client is evidence of a malicious chain. We should have some other primitive defined in x/ibc to deal with malicious chains.
> 
> > A) evidence module is adjusted to handle both malicious-validator evidence and malicious-chain evidence
> 
> I think we should implement a general Evidence interface that only contains the following methods:
> 
> ```go
> // Evidence defines the contract which concrete evidence types of misbehavior
> // must implement.
> type Evidence interface {
> 	Route() string
> 	Type() string
> 	String() string
> 	Hash() cmn.HexBytes
> 	ValidateBasic() error
> }
> ```
> 
> Probably the byzantine validator related methods from the current `Evidence` interface should be abstracted into another interface:
> 
> ```go
>         // The consensus address of the malicious validator at time of infraction
> 	GetConsensusAddress() sdk.ConsAddress
> 
> 	// Height at which the infraction occurred
> 	GetHeight() int64
> 
> 	// The total power of the malicious validator at time of infraction
> 	GetValidatorPower() int64
> 
> 	// The total validator set power at time of infraction
> 	GetTotalPower() int64
> ```
> 
> And ICS02 should implement the Evidence interface from the spec (2 headers) as well as the general interface

This approach seems OK to me. Conceivably other modules might want to handle such misbehaviour as well (e.g. if someone had purchased insurance that pays out if a validator set breaks safety).

Reply:
> Conceivably other modules might want to handle such misbehaviour as well (e.g. if someone had purchased insurance that pays out if a validator set breaks safety).

afaik we haven't implemented slashing logic for this type of misbehaviour (if that's what you're referring with "handle such misbehaviour")